contrib/format-msgpack/src/main/java/org/apache/drill/exec/store/msgpack/MsgpackReader.java

```diff
@@ -0,0 +1,192 @@
+    EnumMap<ValueType, AbstractValueWriter> valueWriterMap = new EnumMap<>(ValueType.class);
+    valueWriterMap.put(ValueType.ARRAY, new ArrayValueWriter(valueWriterMap, emptyArrayWriters));
+    valueWriterMap.put(ValueType.FLOAT, new FloatValueWriter());
+    valueWriterMap.put(ValueType.INTEGER, new IntegerValueWriter());
+    valueWriterMap.put(ValueType.BOOLEAN, new BooleanValueWriter());
+    valueWriterMap.put(ValueType.STRING, new StringValueWriter());
+    valueWriterMap.put(ValueType.BINARY, new BinaryValueWriter());
+    valueWriterMap.put(ValueType.EXTENSION, new ExtensionValueWriter());
+    mapValueWriter = new MapValueWriter(valueWriterMap);
+    valueWriterMap.put(ValueType.MAP, mapValueWriter);
+
+    for (AbstractValueWriter w : valueWriterMap.values()) {
+      w.setup(this.context);
+    }
+  }
+
+  public boolean write(ComplexWriter writer, MaterializedField schema)
+      throws IOException, MessageInsufficientBufferException {
+    if (!unpacker.hasNext()) {
+      return false;
+    }
+
+    context.useSchema = schema != null;
+
+    try {
+      Value v = unpacker.unpackValue();
+      ValueType type = v.getValueType();
+      if (type == ValueType.MAP) {
+        writeRecord(v.asMapValue(), writer, schema);
+      } else {
+        throw new MsgpackParsingException(
+            "Value in root of message pack file is not of type MAP. Skipping type found: " + type);
+      }
+    } catch (MessageInsufficientBufferException e) {
+      throw new MsgpackParsingException("Failed to unpack MAP, possibly because key/value tuples do not match.");
+    }
+    return true;
+  }
+
+  protected void writeRecord(MapValue value, ComplexWriter writer, MaterializedField schema) throws IOException {
+    if (skipQuery) {
+      writer.rootAsMap().bit("count").writeBit(1);
+    } else {
+      mapValueWriter.writeToMap(value, writer.rootAsMap(), this.rootSelection, schema);
+    }
+  }
+
+//
+//
+//  private void ensure(final int length) {
+//    workBuf = workBuf.reallocIfNeeded(length);
+//  }
+
+  @SuppressWarnings("resource")
+  public void ensureAtLeastOneField(ComplexWriter writer) {
+    List<BaseWriter.MapWriter> writerList = Lists.newArrayList();
+    List<PathSegment> fieldPathList = Lists.newArrayList();
+    BitSet emptyStatus = new BitSet(columns.size());
+    int i = 0;
+
+    // first pass: collect which fields are empty
+    for (SchemaPath sp : columns) {
+      PathSegment fieldPath = sp.getRootSegment();
+      BaseWriter.MapWriter fieldWriter = writer.rootAsMap();
+      while (fieldPath.getChild() != null && !fieldPath.getChild().isArray()) {
+        fieldWriter = fieldWriter.map(fieldPath.getNameSegment().getPath());
+        fieldPath = fieldPath.getChild();
+      }
+      writerList.add(fieldWriter);
+      fieldPathList.add(fieldPath);
+      if (fieldWriter.isEmptyMap()) {
+        emptyStatus.set(i, true);
+      }
+      if (i == 0 && !allTextMode) {
```

Comment:
All text mode in JSON means read individual fields, but (since JSON is text), keep them as text to avoid type conflicts.

Not sure it makes sense to do that with MsgPack since fields are typed in the file; you'd have to "un-type" them to convert them to strings.